Foundation/NSLock.swift

```diff
@@ -258,3 +286,80 @@ open class NSCondition: NSObject, NSLocking {
     
     open var name: String?
 }
+
+private func timeSpecFrom(date: Date) -> timespec? {
+    let interval = date.timeIntervalSince1970
+    guard interval >= 0 else {
+        return nil
+    }
+
+    var ts = timespec()
+    ts.tv_sec = Int(interval.rounded(.down))
+    ts.tv_nsec = Int((interval - Double(ts.tv_sec)) * 1_000_000_000.0)
+    return ts
+}
+
+#if os(OSX) || os(iOS)
+
+private func allocateTimedLockData() -> UnsafeMutablePointer<_timed_lock_data>? {
+    let tld = UnsafeMutablePointer<_timed_lock_data>.allocate(capacity: 1)
+    if pthread_mutex_init(&tld.pointee.mutex, nil) == 0 {
+        if pthread_cond_init(&tld.pointee.cond, nil) == 0 {
+            return tld
+        } else {
+            pthread_mutex_destroy(&tld.pointee.mutex)
+        }
+    }
+    tld.deallocate(capacity: 1)
+    return nil
+}
+
+private func deallocateTimedLockData(_ tld: UnsafeMutablePointer<_timed_lock_data>?) {
+    if let tld = tld {
+        pthread_mutex_destroy(&tld.pointee.mutex)
+        pthread_cond_destroy(&tld.pointee.cond)
+        tld.deallocate(capacity: 1)
+    }
+}
+
+private func getOrAllocateTimedLockData(timedLockData: inout UnsafeMutablePointer<_timed_lock_data>?) -> UnsafeMutablePointer<_timed_lock_data> {
+    if timedLockData == nil, let tld = allocateTimedLockData() {
+        var expected: UnsafeMutablePointer<_timed_lock_data>? = nil
+
+        // Use an atomic compare and exchange to set the pointer to the just allocated
+        // data. On X86 this is a LOCK CMPXCHG.
+        // `object` is the target to update, timedLockData, shared by all threads.
+        // `expected` is the value the target should still be (nil).
+        // `desired` is the new value to set it to.
+        if !_stdlib_atomicCompareExchangeStrongPtr(object: &timedLockData,
```

Comment:
oof! are we certain that taking the address here is safe? is it actually atomic?

using _ prefix methods from the standard library should always be done with their consent imho since we are effectively making it API.

@airspeedswift What is your take on using this?

Reply:
I checked the disassembly (on x86_64) and it looked correct. My biggest concern was actually that the documentation seems to imply that the `expect` pointer should be heap allocated but I wasn't sure if that was correct.

I understand your point about _ prefix methods, however we use `@_` function decorators, is that not the same violation? Getting a definitive answer from the Swift core team would be good on both of these points.

I guess as an alternative this could always be done in C in CoreFoundation instead, bypassing these issues.

Reply:
for the specific cases we use `@_` either I have been chastised for using that feature and eventually have been forced to remove it or we have been given special dispensation to utilize it usually with the caveat that when they change things we must change as well.

Reply:
We should definitely not be calling `_stdlib_atomicCompareExchangeStrongPtr` from std lib code.

Reply:
Checking the disassembly is only guaranteeing that it's correct _now_. It could easily change and be incorrect later if it's relying on undefined behavior. But I think (assuming the _ method use is removed) that this usage is correct so long as it is using allocated memory not a property for the `pointee`. cc @jckarter

Reply:
Oops that should read "We should definitely not be calling `_stdlib_atomicCompareExchangeStrongPtr` from *outside* the std lib."

Reply:
@airspeedswift What would you recommend as an alternative?

Reply:
I changed the `_stdlib_atomicCompareExchangeStrongPtr` to using a `lazy var`